connection-service/connection-flask.py

```diff
@@ -7,204 +7,244 @@
-  if part in partitions:
-    store=partitions[part]
-
-    endpoint=json.loads(request.form['endpoint'])
-    matches=[]
-    ctype=""
-    for entry in store['endpoint']:
-      match=True
-      ep=json.loads(entry)
-      for item in endpoint:
-        #print(f"item=<{item}> entry=<{ep[item]}> search=<{endpoint[item]}>")
-        if ep[item]!=endpoint[item] and endpoint[item] != '*':
-          match=False
-      if match:
-        matches.append(store['endpoint'][entry])
-        if entry in store['connection']:
-          conn=store['connection'][entry]
-          ctype=conn['connection_type']
-        if entry in store['connection_type']:
-          ctype=store['connection_type'][entry]
-    if len(matches)>0:
-      result='{'+f'"uris":['
-      for n in range(len(matches)):
-        result=result+'"'+matches[n]+'"'
-        if n<len(matches)-1:
-          result=result+','
-      result=result+'],'+f'"connection_type":"{ctype}"'+'}'
-      #print(f"result = <{result}>")
-      dummy=json.loads(result)
-      print(f"json   = <{result}>")
-      return(result)
-    else:
-      abort(404)
+    good=False
+  partlock.release()
+  if good:
+    return 'OK'
   else:
-    #print(f"Partition {part} not found")
     abort(404)
 
-def lookup(part, map, key):
-  if part in partitions:
-    store=partitions[part]
-    #print(f"Looking for <{key}> in <{store[map]}>")
-    if key in store[map]:
-      val=store[map][key]
-      #print(f"lookup key={key} {map}[key]={val}")
-      return(val)
+@app.route("/getconnection/<part>",methods=['POST','GET'])
+def get_connection(part):
+  # Find connection uris that correspond to the connection id pattern
+  # in the request. The pattern is treated as a regular expression.
+  now=datetime.now()
+  js=json.loads(request.data)
+  if debug_level>2:
+    print (f"[{now}] get_connection() {js=}")
+
+  if 'uid_regex' in js and 'data_type' in js:
+    if debug_level>1:
+      print(f"[{now}] get_connection()"
+            f" Searching for connections matching uid_regex<{js['uid_regex']}>"
+            f" and data_type {js['data_type']}")
+    result=[]
+    regex=re.compile(js['uid_regex'])
+    dt=js['data_type']
+    partlock.acquire()
+    if part in partitions:
+      store=partitions[part]
+      matched=[]
+      for uid,con in store.items():
+        if regex.search(uid) and con.data_type==dt and now-con.time<entry_ttl:
+          #print (f"Found matching entry {uid} {con=}")
+          #result.append('{'
+          #              f'"uid":"{uid}",'
+          #              f'"uri":"{con.uri}",'
+          #              f'"connection_type":{con.connection_type},'
+          #              f'"data_type":"{con.data_type}"'
+          #              '}')
+          matched.append((uid,con))
+      partlock.release()
+      # We should now be able to construct JSON string while other threads
+      # have access to the partition dict
+      for uid,con in matched:
+        result.append('{'
```

Comment:
This could be just an array of dictionaries that gets transformed to json at the end (line 243) with `json.dumps`, but ok.